gnovm/pkg/gnolang/precompile.go

```diff
@@ -246,16 +248,53 @@ func PrecompileBuildPackage(fileOrPkg string, goBinary string) error {
 		cmd.Dir = rootDir
 	}
 	out, err := cmd.CombinedOutput()
-	if err != nil {
-		fmt.Fprintln(os.Stderr, string(out))
-		return fmt.Errorf("std go compiler: %w", err)
+	if _, ok := err.(*exec.ExitError); ok {
+		// exit error
+		return parseGoBuildErrors(string(out))
 	}
+	return err
+}
 
-	return nil
+var errorRe = regexp.MustCompile(`(?m)^(\S+):(\d+):(\d+): (.+)$`)
+
+// parseGoBuildErrors returns a scanner.ErrorList filled with all errors found
+// in out, which is supposed to be the output of the `go build` command.
+// Each errors are translated into their correlated gno files, by:
+// - changing the filename from *.gno.gen.go to *.gno
+// - shifting line number according to the added header in generated go files
+// (see [Precompile] for that header).
+func parseGoBuildErrors(out string) error {
+	var errList goscanner.ErrorList
+	matches := errorRe.FindAllStringSubmatch(out, -1)
+	for _, match := range matches {
+		filename := match[1]
+		line, err := strconv.Atoi(match[2])
+		if err != nil {
+			return fmt.Errorf("parse line go build error %s: %w", match, err)
+		}
+
+		column, err := strconv.Atoi(match[3])
+		if err != nil {
+			return fmt.Errorf("parse column go build error %s: %w", match, err)
+		}
+		msg := match[4]
+		errList.Add(token.Position{
+			// Remove .gen.go extension, we want to target the gno file
+			Filename: strings.TrimSuffix(filename, ".gen.go"),
+			// Shift the 4 lines header added in *.gen.go files.
+			// NOTE(tb): the 4 lines shift below assumes there's always a //go:build
+			// directive. But the tags are optional in the Precompile() function
+			// so that leaves some doubts... We might want something more reliable than
+			// constants to shift lines.
+			Line:   line - 4,
```

Comment:
I'm unfamiliar with how this works.  Why does it get shifted by four lines?

Reply:
@deelawn https://github.com/gnolang/gno/pull/1670/files#diff-69e5bb26203ac2cfe8e34898c91adf9eca6924c6effdafb116bc4d02e9ccd3e5R168-R171